### DISCUSSION TITLE
Attribute initialization table

### DISCUSSION THREAD

**Discussion by colorworlds** [godotengine/godot-proposals#4987] [Editor]
### Problems I encountered

When I used Godot to develop game projects, there was a problem that always bothered me. That is, I found the exported attributes to be unsafe.Because their values are very easy to lose.

For example, I have an exported variable, the equipment name variable:

```
export var equip_name = ''
```
In the editor, I can assign a value to this variable to set different instances, which have the same scene.Like a `sword` and a `helmet`.

But when I use it like this, there will be great risks. These attribute values ,such as `sword` and a `helmet`,may disappear after a certain modification.

There are two common situations that cause attribute values to be lost:

1.Modify attribute name:

I later found that this scene can be applied to all items, not just equipment

```
# export var equip_name = ''
export var item_name = '' 

```
At this time, the attribute values of `equip_name` will disappear. such as `sword` and a `helmet`.

2.Modify the attribute of the parent scene.

If you modify the attribute value of the parent scene, such as setting `equip_name` to an empty string.

the attribute value of the child scene will also become empty string. All attribute values will be lost.

3.There is another additional case. If you localize the scene, you will expand a custom node and assign its attributes, but when you cancel localization, all attribute values will be lost.

In essence, I just want an initial value that can be customized, and the initial value will not be lost casually, so why not use a specific script resource to initialize all variables of a scene?

### Attribute initialization table
I hope to have a table that can initialize the attributes of all nodes in the current scene. I call it `attribute initialization table`.

In fact, it is not just the exported values. When considering inheritance, the values of all child scenes may be lost due to operations in the parent scene.

I think the attribute initialization table should contain two parts:

1.Attribute Name Path (like the track name in the Animation used by the AnimationPlayer , such as `Sprite:position`)

2.Attribute initial value.

### Features
1.`Attribute initialization table` sheet should be a resource.

2.You can specify an `attribute initialization table` for each scene to initialize the scene.

3.`Attribute initialization table` will run  in the editor Similar to the `tool` keyword.It maybe can replace the `tool` keyword.

4.An instance of a scene can be initialized at one time through the code transfer an `Attribute initialization table`. for example:
```
# now
var weapon = Weapon.instance()
weapon.position = Vector2(100,100)
...

# Ideal situation
var weapon_attribute_initialization_table = preload('res://table_path')
var weapon = Weapon.instance(weapon_attribute_initialization_table)

# Possible implementation at present
var weapon = Weapon.instance()
weapon.init_attribute(weapon_attribute_initialization_table)

```
At present, in my game, some child scenes inherit the parent scene just to change the initial value of the parent scene to customize a parent scene. 

If you use an initialization table, you can directly specify different initialization tables for the parent scene to get different scenes, so you don't need to customize the parent scene through inheritance.There is no need to create these sub scenes.
Change logic may be like this:

```
Base Scene <- Sub Scene => for=> custom attribute
Base Scene + Attribute initialization table => for => custom attribute
```

**Comment by Zireael07** [comment]
For exported variables refactoring, I believe there is an open proposal already.

Your proposal seems to boil down to using Resources to initialize values. How is that safe from the same renaming variables problem?

**Reply by colorworlds** [reply]
I feel that this seems to be the same purpose. Isn't the purpose of exporting variables to have a customizable initial value? Or the exported variables can also be used for debugging?  I'm not sure.

If the variable name changes as follows:

```
# export var equip_name = ''
export var item_name = '' 
```
Just change the attribute path:
```
# Suppose the root node of the scene is a Sprite node
# Sprite:equip_name = 'sword'
Sprite:item_name = 'sword'

```
Attribute values `sword` are always saved in the resource.

**Reply by Zireael07** [reply]
You didn't show how the initialization table is implemented. I meant renaming things in that table, not in the sprite itself.

**Reply by colorworlds** [reply]
Sorry, my English is not very good. I use English translation software to communicate.

According to my thinking, the initialization table should be a simplification of `Animation` Resource, and each animation track has only one value.

![截图未命名0727182354](https://user-images.githubusercontent.com/59448002/181223957-33c26cdc-0dae-4874-b5b7-623ac1c28f4c.png)

**Comment by Zireael07** [comment]
IIRC if you rename the variable you're animating, you lose the track (and no, an animation can have multiple values... one track can only animate one variable but can have multiple values, i.e. multiple dots)

**Reply by colorworlds** [reply]
Maybe you have confused the `Animation` and the animation `Track`. Modifying the track name will not lose the attribute value
![截图未命名0727234507](https://user-images.githubusercontent.com/59448002/181289988-b92ee0a9-9288-4872-aac7-c2289b98e9e9.png)

**Reply by Zireael07** [reply]
Oh, didn't know that, that's handy indeed <3
